### DISCUSSION TITLE
blog/how-the-jvm-executes-java-code

### DISCUSSION THREAD

**Discussion by giscus** [cesarsotovalero/cesarsotovalero.github.io#70] [Announcements]
# blog/how-the-jvm-executes-java-code

This blog post describes in details what happens when executing a Java application. After reading this post, you will understand the execution lifecycle of a Java application and the activities performed by the JVM during the execution phase.

https://www.cesarsotovalero.net/blog/how-the-jvm-executes-java-code.html

**Comment by apedano** [comment]
Hi César, first of all thank you for the nice and clear article. One remark though: in the class unloading paragraph you write that the "class or interface" are unloaded, which as far as I know is correct. Your code example, using the WeakReference shows that the "instance" has been garbage collected and not the class itself. I think this is a bit misleading.

**Comment by cesarsotovalero** [comment]
Hi @apedano, thanks for the remark. You're right! The JVM **loads** the binary format of the class (i.e., the `.class` file) and not an instantiated object. The instantiation of objects from the class happens separately and subsequently during the program's execution. On the other hand, the JVM **unloads** a class to free the memory associated with the `Class` object representing that class. After a class has been unloaded, it can be loaded again by another `ClassLoader`, which treats it as a completely new class. In typical standalone applications, class unloading is less common because the system `ClassLoader` is usually active for the lifetime of the application, and hence the classes it loads are not unloaded.

Therefore,  the `LargeClass` is an example of object garbage collection, not class unloading. It demonstrates how a large object can be garbage collected when it is no longer reachable, but this is different from unloading the `LargeClass` class itself. I've clarified this in the article and added an example of a potential class unloading scenario. You can look at the changes here: https://github.com/cesarsotovalero/cesarsotovalero.github.io/commit/783d3f617922f049eea7e834183d4c5136784cf2.

**Comment by luccasalmeida** [comment]
I want to refresh my Java concepts, your article is well structure and clearer.

**Comment by Raxel01** [comment]
I Want to Say Hi ! and at 2026 your article still make impact, Thank you
